src/ports/postgres/modules/stats/pred_metrics.py_in

```diff
@@ -0,0 +1,562 @@
+        help_string = """
+------------------------------------------------------------
+                        USAGE
+------------------------------------------------------------
+SELECT {schema_madlib}.adjusted_r2_score(
+    'table_in',        -- Name of the input table
+    'table_out',       -- Table name to store the metric results
+    'pred_col',        -- Column name containing prediction results
+    'obs_col',         -- Column name containing observed (actual) values
+    'num_predictors',  -- Number of predictors (features) used in the model
+    'training_size',   -- Size of the training data
+    'grouping_cols'    -- Comma-separated list of columns to use as group-by
+);
+    """
+    else:
+        help_string = "No such option. Use {schema_madlib}.adjusted_r2_score('usage')"
+    return help_string.format(**locals())
+
+
+def binary_classifier(table_in, table_out, pred_col, obs_col, grp_col_str=None):
+    """ Get multiple metrics useful for evaluating binary classifiers.
+
+    The 'obs_col' column in 'table_in' is assumed to be a int/float column with
+    two level: 0 and 1, 1 is considered positive and 0 is negative.
+
+    The 'pred_col' is assumed to be a float column that gives the probability of
+    the prediction being positive.
+
+    """
+    grp_cols = _parse_grp_col_str(grp_col_str)
+    _validate_args(table_in, table_out, grp_cols + [pred_col, obs_col])
+    obs_levels = [i['c'] for i in
+                  plpy.execute("SELECT {0} as c FROM {1} GROUP BY {0} ORDER BY {0}".
+                               format(obs_col, table_in))
+                  ]
+    _assert(obs_levels == [0, 1], "Prediction metrics: Observed levels should be 0 and 1")
+
+    if grp_col_str:
+        grp_str = grp_col_str + ","
+        partition_str = "PARTITION BY " + grp_col_str
+    else:
+        grp_str = partition_str = ""
+
+    sql_st = """
+        CREATE TABLE {table_out} AS
+        SELECT *,
+            tp::float8 / NULLIF(tp + fn, 0) AS tpr,
+            tn::float8 / NULLIF(fp + tn, 0) AS tnr,
+            tp::float8 / NULLIF(tp + fp, 0) AS ppv,
+            tn::float8 / NULLIF(tn + fn, 0) AS npv,
+            fp::float8 / NULLIF(fp + tn, 0) AS fpr,
+            fp::float8 / NULLIF(fp + tp, 0) AS fdr,
+            fn::float8 / NULLIF(fn + tp, 0) AS fnr,
+            (tp + tn)::float8 / NULLIF(tp + tn + fp + fn, 0) AS acc,
+            tp * 2.0 / NULLIF(2.0 * tp + fp + fn, 0) AS f1
+        FROM (
+            SELECT
+                {grp_str}
+                threshold,
+                sum(t) OVER ({partition_str} ORDER BY threshold DESC) AS tp,
+                sum(f) OVER ({partition_str} ORDER BY threshold DESC) AS fp,
+                sum(t) OVER ({partition_str}) -
+                    sum(t) OVER ({partition_str} ORDER BY threshold DESC) AS fn,
+                sum(f) OVER ({partition_str}) -
+                    sum(f) OVER ({partition_str} ORDER BY threshold DESC) AS tn
```

Comment:
I suggest defining the window in the query's WINDOW clause, e.g., `WINDOW w AS ({partition_str} ORDER BY threshold DESC)` and just reference it in the query, e.g., `sum(t) over w AS tp`.